src/org/opendatakit/briefcase/util/FormCache.java

```diff
@@ -1,97 +1,149 @@
+import static java.util.stream.Collectors.toList;
+import static org.opendatakit.briefcase.reused.UncheckedFiles.createFile;
+import static org.opendatakit.briefcase.reused.UncheckedFiles.delete;
+import static org.opendatakit.briefcase.reused.UncheckedFiles.list;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.InvalidClassException;
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
 import java.io.OutputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
+import org.bushe.swing.event.EventBus;
+import org.bushe.swing.event.annotation.AnnotationProcessor;
+import org.bushe.swing.event.annotation.EventSubscriber;
 import org.opendatakit.briefcase.model.BriefcaseFormDefinition;
-import org.opendatakit.briefcase.reused.BriefcaseException;
-import org.opendatakit.briefcase.reused.UncheckedFiles;
+import org.opendatakit.briefcase.pull.PullEvent;
+import org.opendatakit.briefcase.reused.CacheUpdateEvent;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class FormCache implements FormCacheable {
+public class FormCache {
   private static final Logger log = LoggerFactory.getLogger(FormCache.class);
   private static final String CACHE_FILE_NAME = "cache.ser";
-  private final Path cacheFile;
-  private final Map<String, String> pathToMd5Map;
-  private final Map<String, BriefcaseFormDefinition> pathToDefinitionMap;
+  private Optional<Path> cacheFile;
+  private Optional<Path> briefcaseDir;
+  private Map<String, String> hashByPath;
+  private Map<String, BriefcaseFormDefinition> formDefByPath;
 
-  public FormCache(Path cacheFile, Map<String, String> pathToMd5Map, Map<String, BriefcaseFormDefinition> pathToDefinitionMap) {
+  private FormCache(Optional<Path> cacheFile, Map<String, String> hashByPath, Map<String, BriefcaseFormDefinition> formDefByPath) {
     this.cacheFile = cacheFile;
-    this.pathToMd5Map = pathToMd5Map;
-    this.pathToDefinitionMap = pathToDefinitionMap;
-    Runtime.getRuntime().addShutdownHook(new Thread(this::save));
+    this.briefcaseDir = cacheFile.map(Path::getParent);
+    this.hashByPath = hashByPath;
+    this.formDefByPath = formDefByPath;
+    AnnotationProcessor.process(this);
+  }
+
+  public static FormCache empty() {
+    return new FormCache(Optional.empty(), new HashMap<>(), new HashMap<>());
   }
 
-  @SuppressWarnings("unchecked")
   public static FormCache from(Path briefcaseDir) {
-    Path cacheFile = briefcaseDir.resolve(CACHE_FILE_NAME);
-    if (exists(cacheFile))
-      try (InputStream in = Files.newInputStream(cacheFile);
+    FormCache formCache = empty();
+    formCache.setLocation(briefcaseDir);
+    return formCache;
+  }
+
+  @SuppressWarnings("unchecked")
+  public void setLocation(Path newBriefcaseDir) {
```

Comment:
It looks like now when setting a location to an existing directory with a valid cache, that cache is rebuilt. Is that intended?

Reply:
Yes, we can't be sure that the user hasn't changed the files inside. The only time we don't  want to rebuild it is on startup, which, now that you mention, I think I need to review